util/matchers.go

```diff
@@ -30,19 +30,19 @@ func ExtractMetricNameFromMetric(m model.Metric) (model.LabelValue, error) {
 }
 
 // ExtractMetricNameFromMatchers extracts the metric name from a set of matchers
-func ExtractMetricNameFromMatchers(matchers []*metric.LabelMatcher) (model.LabelValue, []*metric.LabelMatcher, error) {
+func ExtractMetricNameFromMatchers(matchers []*metric.LabelMatcher) (model.LabelValue, []*metric.LabelMatcher, bool, error) {
 	outMatchers := make([]*metric.LabelMatcher, len(matchers)-1)
 	for i, matcher := range matchers {
 		if matcher.Name != model.MetricNameLabel {
 			continue
 		}
 		if matcher.Type != metric.Equal {
-			return "", nil, fmt.Errorf("must have equality matcher for MetricNameLabel")
+			return "", nil, false, fmt.Errorf("must have equality matcher for MetricNameLabel")
```

Comment:
I would remove this error and just return false, as in this case of a non equality matcher (ie a regex, or a !=), we just need to send the request to all ingesters again.

In fact, in the case of non equality, we can do some optimisations on the chunk store too.